instruments/abstract_instruments/comm/cryomagnetics_serial_communicator.py

```diff
@@ -0,0 +1,167 @@
+            push the response length over 140 characters. In that case, the
+            maximum message size may need to be set higher
+
+        .. todo::
+
+            Investigate a batching system to place multiple commands sent
+            within a short timespan into a single command, and return the
+            responses appropriately. The query method would then likely be
+            based on a priority queue
+
+        :return: The maximum message size in characters
+        :rtype: int
+        """
+        return self._maximum_message_size
+
+    # METHODS #
+
+    @maximum_message_size.setter
+    def maximum_message_size(self, new_size):
+        """
+        Set the maximum message size that will be read out by the communicator
+
+        :param int new_size: The new size to be read out
+        """
+        self._maximum_message_size = new_size
+
+    def seek(self, offset):
+        """
+        Go to a particular point in the data stream. This is not implemented
+        in this communicator
+
+        :param offset: The amount of bytes to fast forward
+        """
+        raise NotImplementedError
+
+    def tell(self):
+        """
+        Get the current positional offset for the data stream. This is not
+        implemented in this communicator
+        """
+        raise NotImplementedError
+
+    def _query(self, msg, size=-1):
+        """
+        This is the implementation of ``query`` for communicating with
+        Cryomagnetics instruments over serial connections. This function is,
+        in turn, wrapped by the concrete method `AbstractCommunicator.query` to
+        provide consistent logging functionality across all communication
+        layers.
+
+        :param str msg: The query message to send to the instrument
+        :param int size: The number of bytes to read back from the
+            instrument response. If the size is -1, the size will be assumed
+            to be equal to
+            `CryomagneticsSerialCommunicator.maximum_message_size`
+        :return: The instrument response to the query
+        :rtype: `str`
+        """
+        size_to_read = self._get_required_message_size(size)
+
+        self._querying_lock.acquire()
+        self.terminator = '\r'
+        self.write(msg + self.terminator)
+        self.terminator = '\r\n'
+        response = self.read(size=size_to_read)
+        self._querying_lock.release()
+
+        return self._parse_query(msg, response)
+
+    def _get_required_message_size(self, size):
+        """
+        If the size is not negative, return the message size. Otherwise,
+        return the current maximum message size
+
+        :param int size: The size value to parse
+        :return: The number of bytes that will be read out of the serial port
+        :rtype: int
+        """
+        if size < 0:
+            return self.maximum_message_size
+        else:
+            return size
+
+    def _parse_query(self, command, response):
```

Comment:
Out of curiousity, is it possible to separate out the logic for communicators which handle echoes? Perhaps even more generally, communicators which handle postprocessing? In particular, that could help prevent bugs like ``query()``/``sendcmd()`` being introduced with a change in communicator class.